File: C#/Silverlight/source/LSilverlight-0.3.3/Loon.Core.Input/LTransition.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Loon.Core.Graphics;
using Loon.Utils;
using Loon.Core.Graphics.OpenGL;
using Loon.Action.Sprite.Effect;
using Loon.Action.Map;
using Loon.Action.Sprite;

namespace Loon.Core.Input
{
    public class LTransition
    {

        public interface TransitionListener
        {
            void Update(long elapsedTime);

            void Draw(GLEx g);

            bool Completed();

            void Dispose();
        }

        public static LTransition NewCrossRandom()
        {
            return NewCrossRandom(LColor.black);
        }

        public static LTransition NewCrossRandom(LColor c)
        {
            return NewCross(MathUtils.Random(0, 1), TextureUtils
                    .CreateTexture(LSystem.screenRect.width,
                            LSystem.screenRect.height, c));
        }

        public static LTransition NewCross(int c)
        {
            return NewCross(c, ScreenUtils.ToScreenCaptureTexture());
        }

        private class _Cross : TransitionListener
        {

            CrossEffect cross;

            public _Cross(int c, LTexture texture)
            {
                cross = new CrossEffect(c, texture);
            }

            public void Draw(GLEx g)
            {
                cross.CreateUI(g);
            }

            public void Update(long elapsedTime)
            {
                cross.Update(elapsedTime);
            }

            public bool Completed()
            {
                return cross.IsComplete();
            }

            public void Dispose()
            {
                cross.Dispose();
            }

        }



        public static LTransition NewCross(int c, LTexture texture)
        {
            if (GLEx.Self != null)
            {
                LTransition transition = new LTransition();
                transition.SetTransitionListener(new _Cross(c, texture));
                transition.SetDisplayGameUI(true);
                transition.code = 1;
                return transition;
            }
            return null;
        }

        private class _Arc : TransitionListener
        {

            ArcEffect arc;

            public _Arc(LColor c)
            {
                arc = new ArcEffect(c);
            }

            public void Draw(GLEx g)
            {
                arc.CreateUI(g);
            }

            public void Update(long elapsedTime)
            {
                arc.Update(elapsedTime);
            }

            public bool Completed()
            {
                return arc.IsComplete();
            }

            public void Dispose()
            {
                arc.Dispose();
            }
        }

        public static LTransition NewArc(LColor c)
        {
            if (GLEx.Self != null)
            {

                LTransition transition = new LTransition();
                transition.SetTransitionListener(new _Arc(c));
                transition.SetDisplayGameUI(true);
                transition.code = 1;
                return transition;
            }
            return null;
        }

        public static LTransition NewArc()
        {
            return NewArc(LColor.black);
        }

        private class _Split : TransitionListener
        {

            SplitEffect split;

            public _Split(int d, LTexture texture)
            {
                split = new SplitEffect(texture, d);
            }

            public void Draw(GLEx g)
            {
                split.CreateUI(g);
            }

            public void Update(long elapsedTime)
            {
                split.Update(elapsedTime);
            }

            public bool Completed()
            {
                return split.IsComplete();
            }

            public void Dispose()
            {
                split.Dispose();
            }

        }

        public static LTransition NewSplit(int d)
        {
            return NewSplit(d, ScreenUtils.ToScreenCaptureTexture());
        }

        public static LTransition NewSplit(int d, LTexture texture)
        {
            if (GLEx.Self != null)
            {

                LTransition transition = new LTransition();
                transition.SetTransitionListener(new _Split(d, texture));
                transition.SetDisplayGameUI(true);
                transition.code = 1;
                return transition;
            }
            return null;
        }

        public static LTransition NewSplitRandom()
        {
            return NewSplit(MathUtils.Random(0, Config.TDOWN));
        }

        public static LTransition NewSplitRandom(LTexture texture)
        {
            return NewSplit(MathUtils.Random(0, Config.TDOWN), texture);
        }

        public static LTransition NewSplitRandom(LColor c)
        {
            return NewSplitRandom(TextureUtils.CreateTexture(
                    LSystem.screenRect.width, LSystem.screenRect.height, c));
        }


        private class _Out : TransitionListener
        {

            OutEffect split;

            public _Out(int d, LTexture texture)
            {
                split = new OutEffect(texture, d);
            }

            public void Draw(GLEx g)
            {
                split.CreateUI(g);
            }

            public void Update(long elapsedTime)
            {
                split.Update(elapsedTime);
            }

            public bool Completed()
            {
                return split.IsComplete();
            }

            public void Dispose()
            {
                split.Dispose();
            }

        }

        public static LTransition NewOut(int d, LTexture texture)
        {
            if (GLEx.Self != null)
            {

                LTransition transition = new LTransition();
                transition.SetTransitionListener(new _Out(d, texture));
                transition.SetDisplayGameUI(true);
                transition.code = 1;
                return transition;
            }
            return null;
        }

        private class _Fade : TransitionListener
        {

            FadeEffect fade;

            public _Fade(int type, LColor c)
            {
                fade = FadeEffect.GetInstance(type, c);
            }

            public void Draw(GLEx g)
            {
                fade.CreateUI(g);
            }

            public void Update(long elapsedTime)
            {
                fade.Update(elapsedTime);
            }

            public bool Completed()
            {
                return fade.IsStop();
            }

            public void Dispose()
            {
                fade.Dispose();
            }

        }

        public static LTransition NewFade(int type, LColor c)
        {
            if (GLEx.Self != null)
            {
                LTransition transition = new LTransition();
                transition.SetTransitionListener(new _Fade(type, c));
                transition.SetDisplayGameUI(true);
                transition.code = 1;
                return transition;
            }
            return null;
        }

        public static LTransition NewFadeIn()
        {
            return LTransition.NewFade(FadeEffect.TYPE_FADE_IN);
        }

        public static LTransition NewFadeOut()
        {
            return LTransition.NewFade(FadeEffect.TYPE_FADE_OUT);
        }

        public static LTransition NewFade(int type)
        {
            return LTransition.NewFade(type, LColor.black);
        }


        private class _Cycle : TransitionListener
        {

            WaitSprite wait;

            public _Cycle(int d)
            {
                wait = new WaitSprite(d);
            }

            public void Draw(GLEx g)
            {
                wait.CreateUI(g);
            }

            public void Update(long elapsedTime)
            {
                wait.Update(elapsedTime);
            }

            public bool Completed()
            {
                return true;
            }

            public void Dispose()
            {
                wait.Dispose();
            }

        }

        public static LTransition NewCycle(int type)
        {
            if (GLEx.Self != null)
            {
                LTransition transition = new LTransition();
                transition.SetTransitionListener(new _Cycle(type));
                transition.SetDisplayGameUI(false);
                transition.code = 0;
                return transition;
            }
            return null;
        }

        public static LTransition NewCycleRandom()
        {
            return NewCycle(MathUtils.Random(0, 11));
        }

        private class _PShadow : TransitionListener
        {

            PShadowEffect effect;

            public _PShadow(PShadowEffect e)
            {
                effect = e;
            }

            public void Draw(GLEx g)
            {
                effect.CreateUI(g);
            }

            public void Update(long elapsedTime)
            {
                effect.Update(elapsedTime);
            }

            public bool Completed()
            {
                return effect.IsComplete();
            }

            public void Dispose()
            {
                effect.Dispose();
            }

        }

        public static LTransition NewPShadow(string fileName, float alhpa)
        {
            PShadowEffect shadow = new PShadowEffect(fileName);
            shadow.SetAlpha(alhpa);
            return NewPShadow(shadow);
        }

        public static LTransition NewPShadow(string fileName)
        {
            return NewPShadow(fileName, 0.5f);
        }

        public static LTransition newPShadowGameUI(string fileName)
        {
            return NewPShadow(PShadowEffect.NewScreenEffect(fileName));
        }

        public static LTransition NewPShadow(PShadowEffect effect)
        {
            if (GLEx.Self != null)
            {
                LTransition transition = new LTransition();
                transition.SetTransitionListener(new _PShadow(effect));
                transition.SetDisplayGameUI(true);
                transition.code = 1;
                return transition;
            }
            return null;
        }

        private class _Empty : TransitionListener
        {


            public _Empty()
            {

            }

            public void Draw(GLEx g)
            {

            }

            public void Update(long elapsedTime)
            {

            }

            public bool Completed()
            {
                return true;
            }

            public void Dispose()
            {

            }

        }


        public static LTransition NewEmpty()
        {
            if (GLEx.Self != null)
            {
                LTransition transition = new LTransition();
                transition.SetTransitionListener(new _Empty());
                transition.SetDisplayGameUI(false);
                transition.code = 0;
                return transition;
            }
            return null;
        }

        // 是否在在启动过渡效果同时显示游戏画面（即是否顶层绘制过渡画面，底层同时绘制标准游戏画面）
        internal bool isDisplayGameUI;

        internal int code;

        internal TransitionListener listener;

        public void SetDisplayGameUI(bool s)
        {
            this.isDisplayGameUI = s;
        }

        public bool IsDisplayGameUI()
        {
            return this.isDisplayGameUI;
        }

        public void SetTransitionListener(TransitionListener l)
        {
            this.listener = l;
        }

        public TransitionListener GetTransitionListener()
        {
            return this.listener;
        }

        internal void Update(long elapsedTime)
        {
            if (listener != null)
            {
                listener.Update(elapsedTime);
            }
        }

        internal void Draw(GLEx g)
        {
            if (listener != null)
            {
                listener.Draw(g);
            }
        }

        internal bool Completed()
        {
            if (listener != null)
            {
                return listener.Completed();
            }
            return false;
        }

        internal void Dispose()
        {
            if (listener != null)
            {
                listener.Dispose();
            }
        }
    }
}

File: Java/Android/OpenGLES/source/core/src/loon/core/input/LTransition.java
package loon.core.input;

import loon.action.map.Config;
import loon.action.sprite.ISprite;
import loon.action.sprite.effect.ArcEffect;
import loon.action.sprite.effect.CrossEffect;
import loon.action.sprite.effect.FadeEffect;
import loon.action.sprite.effect.PShadowEffect;
import loon.action.sprite.effect.SplitEffect;
import loon.core.LSystem;
import loon.core.graphics.LColor;
import loon.core.graphics.opengl.GLEx;
import loon.core.graphics.opengl.LTexture;
import loon.core.graphics.opengl.TextureUtils;
import loon.utils.MathUtils;


/**
 * Copyright 2008 - 2011
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */
/**
 * 自0.3.2版起新增的Screen切换过渡效果类，内置有多种过渡特效。
 * 
 * example:
 * 
 * public class Sample extends Screen{
 * 
 * ......
 * 
 * public LTransition onTransition(){ return LTransition.xxx(method) } }
 * 
 * 
 */
public class LTransition {

	/**
	 * 随机的百叶窗特效
	 * 
	 * @return
	 */
	public final static LTransition newCrossRandom() {
		return newCrossRandom(LColor.black);
	}

	/**
	 * 百叶窗特效
	 * 
	 * @param c
	 * @return
	 */
	public final static LTransition newCrossRandom(LColor c) {
		return newCross(MathUtils.random(0, 1), TextureUtils
				.createTexture(LSystem.screenRect.width,
						LSystem.screenRect.height, c));
	}


	/**
	 * 百叶窗特效
	 * 
	 * @param c
	 * @return
	 */
	public final static LTransition newCross(final int c, final LTexture texture) {

		if (GLEx.self != null) {

			final LTransition transition = new LTransition();

			transition.setTransitionListener(new TransitionListener() {

				final CrossEffect cross = new CrossEffect(c, texture);

				@Override
				public void draw(GLEx g) {
					cross.createUI(g);
				}

				@Override
				public void update(long elapsedTime) {
					cross.update(elapsedTime);
				}

				@Override
				public boolean completed() {
					return cross.isComplete();
				}

				@Override
				public void dispose() {
					cross.dispose();
				}

			});
			transition.setDisplayGameUI(true);
			transition.code = 1;
			return transition;
		}
		return null;
	}


	/**
	 * 默认使用黑色的圆弧渐变特效
	 * 
	 * @return
	 */
	public final static LTransition newArc() {
		return newArc(LColor.black);
	}

	/**
	 * 单一色彩的圆弧渐变特效
	 * 
	 * @return
	 */
	public final static LTransition newArc(final LColor c) {

		if (GLEx.self != null) {

			final LTransition transition = new LTransition();

			transition.setTransitionListener(new TransitionListener() {

				final ArcEffect arc = new ArcEffect(c);

				@Override
				public void draw(GLEx g) {
					arc.createUI(g);
				}

				@Override
				public void update(long elapsedTime) {
					arc.update(elapsedTime);
				}

				@Override
				public boolean completed() {
					return arc.isComplete();
				}

				@Override
				public void dispose() {
					arc.dispose();
				}

			});
			transition.setDisplayGameUI(true);
			transition.code = 1;
			return transition;
		}
		return null;
	}


	/**
	 * 产生一个Screen画面向双向分裂的过渡特效
	 * 
	 * @param texture
	 * @return
	 */
	public final static LTransition newSplitRandom(LTexture texture) {
		return newSplit(MathUtils.random(0, Config.TDOWN), texture);
	}

	/**
	 * 产生一个Screen画面向双向分裂的过渡特效
	 * 
	 * @param c
	 * @return
	 */
	public final static LTransition newSplitRandom(LColor c) {
		return newSplitRandom(TextureUtils.createTexture(
				LSystem.screenRect.width, LSystem.screenRect.height, c));
	}


	/**
	 * 产生一个Screen画面向双向分裂的过渡特效(方向的静态值位于Config类中)
	 * 
	 * @param d
	 * @param texture
	 * @return
	 */
	public final static LTransition newSplit(final int d, final LTexture texture) {

		if (GLEx.self != null) {

			final LTransition transition = new LTransition();

			transition.setTransitionListener(new TransitionListener() {

				final SplitEffect split = new SplitEffect(texture, d);

				@Override
				public void draw(GLEx g) {
					split.createUI(g);
				}

				@Override
				public void update(long elapsedTime) {
					split.update(elapsedTime);
				}

				@Override
				public boolean completed() {
					return split.isComplete();
				}

				@Override
				public void dispose() {
					split.dispose();
				}

			});
			transition.setDisplayGameUI(true);
			transition.code = 1;
			return transition;
		}
		return null;
	}

	/**
	 * 产生一个黑色的淡入效果
	 * 
	 * @return
	 */
	public final static LTransition newFadeIn() {
		return LTransition.newFade(ISprite.TYPE_FADE_IN);
	}

	/**
	 * 产生一个黑色的淡出效果
	 * 
	 * @return
	 */
	public final static LTransition newFadeOut() {
		return LTransition.newFade(ISprite.TYPE_FADE_OUT);
	}

	/**
	 * 产生一个黑色的淡入/淡出效果
	 * 
	 * @param type
	 * @return
	 */
	public final static LTransition newFade(int type) {
		return LTransition.newFade(type, LColor.black);
	}

	/**
	 * 产生一个指定色彩的淡入效果
	 * 
	 * @param c
	 * @return
	 */
	public final static LTransition newFade(final int type, final LColor c) {
		if (GLEx.self != null) {
			final LTransition transition = new LTransition();

			transition.setTransitionListener(new TransitionListener() {

				final FadeEffect fade = FadeEffect.getInstance(type, c);

				@Override
				public void draw(GLEx g) {
					fade.createUI(g);
				}

				@Override
				public void update(long elapsedTime) {
					fade.update(elapsedTime);
				}

				@Override
				public boolean completed() {
					return fade.isStop();
				}

				@Override
				public void dispose() {
					fade.dispose();
				}

			});
			transition.setDisplayGameUI(true);
			transition.code = 1;
			return transition;
		}
		return null;
	}

	/**
	 * 以指定地址图片作渐变式过滤(实际上这种过渡方式所能产生的效果是无穷的，具体可参考吉里吉里(krkr) 用过渡图)
	 * 
	 * @param fileName
	 * @param alhpa
	 */
	public final static LTransition newPShadow(String fileName, float alhpa) {
		PShadowEffect shadow = new PShadowEffect(fileName);
		shadow.setAlpha(alhpa);
		return newPShadow(shadow);
	}

	/**
	 * 以指定地址图片作渐变式过滤
	 * 
	 * @param fileName
	 */
	public final static LTransition newPShadow(String fileName) {
		return newPShadow(fileName, 0.5f);
	}

	/**
	 * 转化PShadowEffect为LTransition
	 * 
	 * @param effect
	 * @return
	 */
	public final static LTransition newPShadow(final PShadowEffect effect) {
		if (GLEx.self != null) {

			final LTransition transition = new LTransition();

			transition.setTransitionListener(new TransitionListener() {

				@Override
				public void draw(GLEx g) {
					effect.createUI(g);
				}

				@Override
				public void update(long elapsedTime) {
					effect.update(elapsedTime);
				}

				@Override
				public boolean completed() {
					return effect.isComplete();
				}

				@Override
				public void dispose() {
					effect.dispose();
				}

			});
			transition.setDisplayGameUI(true);
			transition.code = 1;
			return transition;
		}
		return null;
	}

	public final static LTransition newEmpty() {

		final LTransition transition = new LTransition();

		transition.setTransitionListener(new TransitionListener() {

			@Override
			public void draw(GLEx g) {
			}

			@Override
			public void update(long elapsedTime) {
			}

			@Override
			public boolean completed() {
				return true;
			}

			@Override
			public void dispose() {
			}

		});

		transition.setDisplayGameUI(true);
		transition.code = 1;
		return transition;

	}

	public static interface TransitionListener {

		public void update(long elapsedTime);

		public void draw(GLEx g);

		public boolean completed();

		public void dispose();
	}

	// 是否在在启动过渡效果同时显示游戏画面（即是否顶层绘制过渡画面，底层同时绘制标准游戏画面）
	boolean isDisplayGameUI;

	int code;

	TransitionListener listener;

	public void setDisplayGameUI(boolean s) {
		this.isDisplayGameUI = s;
	}

	public boolean isDisplayGameUI() {
		return this.isDisplayGameUI;
	}

	public void setTransitionListener(TransitionListener l) {
		this.listener = l;
	}

	public TransitionListener getTransitionListener() {
		return this.listener;
	}

	final void update(long elapsedTime) {
		if (listener != null) {
			listener.update(elapsedTime);
		}
	}

	final void draw(GLEx g) {
		if (listener != null) {
			listener.draw(g);
		}
	}

	final boolean completed() {
		if (listener != null) {
			return listener.completed();
		}
		return false;
	}

	final void dispose() {
		if (listener != null) {
			listener.dispose();
		}
	}
}
